tests/scripts/all.sh

```diff
@@ -3992,6 +3992,21 @@ component_test_tls13_only_ephemeral () {
     tests/ssl-opt.sh
 }
 
+component_test_tls13_only_ephemeral_ffdh () {
+    msg "build: TLS 1.3 only from default, only ephemeral ffdh key exchange mode"
+    scripts/config.py unset MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_ENABLED
+    scripts/config.py unset MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_PSK_EPHEMERAL_ENABLED
+    scripts/config.py unset MBEDTLS_SSL_EARLY_DATA
+    scripts/config.py -f include/psa/crypto_config.h unset PSA_WANT_ALG_ECDH
```

Comment:
You must unset MBEDTLS_ECDH_C otherwise this does not have much effect (ECDHE TLS 1.3 tests still run successfully) and then it does not seem to compile.
Otherwise, what about to have the ECDHE equivalent?

Reply:
Ok, because `MBEDTLS_ECDH_C` re-enables `PSA_WANT_ALG_ECDH` in `config_psa.h`.
Now when disabled `MBEDTLS_ECDH_C` check of configuration has failed:
`#error "MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED defined, but not all prerequisites"`

New version should be:
```
 #if defined(MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED)
-#if !( defined(PSA_WANT_ALG_ECDH) && defined(MBEDTLS_X509_CRT_PARSE_C) && \
+#if !( (defined(PSA_WANT_ALG_ECDH) || defined(PSA_WANT_ALG_FFDH)) && defined(MBEDTLS_X509_CRT_PARSE_C) && \
        ( defined(MBEDTLS_PK_HAVE_ECDSA) || defined(MBEDTLS_PKCS1_V21) ) )
 #error "MBEDTLS_SSL_TLS1_3_KEY_EXCHANGE_MODE_EPHEMERAL_ENABLED defined, but not all prerequisites"
```

With this change library builds successfully, but ssl test suite and ``ssl-opt` tests are failing. Will continue investigation tomorrow.

Reply:
I check few failing `ssl-opt` tests and I'm confused:
```
przemek@przemek-VirtualBox:~/mbedtls$ tests/ssl-opt.sh -f "keyUsage cli 1.3: DigitalSignature, ECDSA: OK"
keyUsage cli 1.3: DigitalSignature, ECDSA: OK .......................... FAIL
  ! bad client exit code (expected 0, got 1)
  ! outputs saved to o-XXX-1.log
------------------------------------------------------------------------
```

In configuration when `ECDH` is disabled client hello message consists ECDH groups:
```
ssl_client.c:0242: |3| client hello, adding supported_groups extension
ssl_client.c:0259: |1| got supported group(001d)
ssl_client.c:0273: |3| NamedGroup: x25519 ( 1d )
ssl_client.c:0259: |1| got supported group(0017)
ssl_client.c:0273: |3| NamedGroup: secp256r1 ( 17 )
ssl_client.c:0259: |1| got supported group(0018)
ssl_client.c:0273: |3| NamedGroup: secp384r1 ( 18 )
ssl_client.c:0259: |1| got supported group(001e)
ssl_client.c:0273: |3| NamedGroup: x448 ( 1e )
ssl_client.c:0259: |1| got supported group(0019)
ssl_client.c:0273: |3| NamedGroup: secp521r1 ( 19 )
ssl_client.c:0259: |1| got supported group(001a)
ssl_client.c:0259: |1| got supported group(001b)
ssl_client.c:0259: |1| got supported group(001c)
ssl_client.c:0259: |1| got supported group(0100)
ssl_client.c:0282: |3| NamedGroup: ffdhe2048 ( 100 )
ssl_client.c:0259: |1| got supported group(0101)
ssl_client.c:0282: |3| NamedGroup: ffdhe3072 ( 101 )
ssl_client.c:0259: |1| got supported group(0102)
ssl_client.c:0282: |3| NamedGroup: ffdhe4096 ( 102 )
ssl_client.c:0259: |1| got supported group(0103)
ssl_client.c:0282: |3| NamedGroup: ffdhe6144 ( 103 )
ssl_client.c:0259: |1| got supported group(0104)
ssl_client.c:0282: |3| NamedGroup: ffdhe8192 ( 104 )
```

Then group 29 is selected and I think this should not happen in this configuration:
```
ssl_tls13_client.c:1502: |2| received HelloRetryRequest message
ssl_tls13_client.c:1617: |3| dumping 'server hello, version' (2 bytes)
ssl_tls13_client.c:1617: |3| 0000:  03 03                                            ..
ssl_tls13_client.c:1571: |3| dumping 'Session ID' (32 bytes)
ssl_tls13_client.c:1571: |3| 0000:  95 c5 29 4a 8f 68 ca 72 7e f3 e3 83 ec 13 fe 7d  ..)J.h.r~......}
ssl_tls13_client.c:1571: |3| 0010:  dd 76 3f 0f 5a 46 ea e7 c8 91 9b 4f ca 6c dc cd  .v?.ZF.....O.l..
ssl_tls13_client.c:1700: |3| server hello, chosen ciphersuite: ( 1303 ) - TLS1-3-CHACHA20-POLY1305-SHA256
ssl_tls13_client.c:1735: |3| dumping 'server hello extensions' (12 bytes)
ssl_tls13_client.c:1735: |3| 0000:  00 2b 00 02 03 04 00 33 00 02 00 1d              .+.....3....
ssl_tls13_generic.c:1638: |3| HelloRetryRequest: supported_versions(43) extension received.
ssl_tls13_generic.c:1638: |3| HelloRetryRequest: key_share(51) extension received.
ssl_tls13_client.c:1797: |3| found key_shares extension
ssl_tls13_client.c:0395: |3| dumping 'key_share extension' (2 bytes)
ssl_tls13_client.c:0395: |3| 0000:  00 1d                                            ..
ssl_tls13_client.c:0400: |3| selected_group ( 29 )
```

As a result computation fails:
`ssl_tls13_generic.c:1562: |1| Perform PSA-based ECDH/FFDH computation.`

I don't know if this is component configuration error, test configuration error or problem in code. @mpg @ronald-cron-arm Can you please give some hint.

I'm also not sure if this part is correct. For ECDH we have `*group_list != selected_group` condition, but I don't understan what it stands for:
https://github.com/Mbed-TLS/mbedtls/pull/7627/files#diff-ef6f5faac0e0b98311d2daf2dfad61a856336e27a7a22ee248b88e72013781efR411-R426

Why ar

Reply:
Regarding the EC groups in supported_groups extension, I can see that even if ECDH is disabled EC groups are in the default[ list of groups] (https://github.com/Mbed-TLS/mbedtls/blob/6edf8b8c7beeac4bcf27671d1afe37716a898814/library/ssl_tls.c#LL4928C35-L4928C35) and in `ssl_write_supported_groups_ext()`  the code to propose EC groups is guarded by MBEDTLS_ECP_LIGHT, not PSA_WANT_ALG_ECDH. I do no see why it is not guarded by PSA_WANT_ALG_ECDH at least for TLS 1.3.

Reply:
Yes the group are in the default gropus and one of options to solve this is to use `PSA_WANT_ALG_ECDH` guard in  `ssl_write_supported_groups_ext`:
```
#if defined(PSA_WANT_ALG_ECDH)
        if ((mbedtls_ssl_conf_is_tls13_enabled(ssl->conf) &&
             mbedtls_ssl_tls13_named_group_is_ecdhe(*group_list)) ||
            (mbedtls_ssl_conf_is_tls12_enabled(ssl->conf) &&
             mbedtls_ssl_tls12_named_group_is_ecdhe(*group_list))) {
            if (mbedtls_ssl_get_ecp_group_id_from_tls_id(*group_list) ==
                MBEDTLS_ECP_DP_NONE) {
                continue;
            }
            MBEDTLS_SSL_CHK_BUF_PTR(p, end, 2);
            MBEDTLS_PUT_UINT16_BE(*group_list, p, 0);
            p += 2;
            MBEDTLS_SSL_DEBUG_MSG(3, ("NamedGroup: %s ( %x )",
                                      mbedtls_ssl_get_curve_name_from_tls_id(*group_list),
                                      *group_list));
        }
#endif /* PSA_WANT_ALG_ECDH */
```

Reply:
The fact that EC groups are in the default list of groups even if ECDH is disabled is probably wrong as well but the list of groups can also be set-up through `mbedtls_ssl_conf_groups` thus better to make the change in `ssl_write_supported_groups_ext` anyway. I am not completely sure about the TLS 1.2 case though. In TLS 1.2 can a group/curve be proposed for something else than ECDH? @mpg would you know about that?

Reply:
> In TLS 1.2 can a group/curve be proposed for something else than ECDH? @mpg would you know about that?

Given the guards of `ssl_write_supported_groups_ext()` (#if defined(MBEDTLS_ECDH_C) || defined(MBEDTLS_ECDSA_C) || \                    
    defined(MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED) || defined(PSA_WANT_ALG_FFDH)) it is likely to be the case. But if TLS 1.3 is proposed we should probably not include the EC groups if ECDH is not supported as we end up declaring that we can do something but in fact we cannot.

Reply:
Yes it seems first that we are missing a lot of dependencies on ECDH.

Reply:
Lots of generated compat-tls13 tests are failing so I would start with adding `requires_config_enabled MBEDTLS_ECDH_C` to test case that are using mbedtls and EC groups.